packages/aws-cdk/lib/api/logs/find-cloudwatch-logs.ts

```diff
@@ -0,0 +1,108 @@
+  // TODO: update this to use the lookup role once https://github.com/aws/aws-cdk/pull/18277 is merged
+  const sdk = await sdkProvider.forEnvironment(resolvedEnv, Mode.ForReading);
+
+  const listStackResources = new LazyListStackResources(sdk, stackArtifact.stackName);
+  const evaluateCfnTemplate = new EvaluateCloudFormationTemplate({
+    stackArtifact,
+    parameters: {},
+    account: resolvedEnv.account,
+    region: resolvedEnv.region,
+    partition: (await sdk.currentAccount()).partition,
+    urlSuffix: sdk.getEndpointSuffix,
+    listStackResources,
+  });
+
+  // map of logicalId to CloudFormation type
+  // e.g. 'mylambdaFunction': 'AWS::Lambda::Function'
+  const logicalIdsOfImplicitLogServices = new Map<string, string>();
+  const logGroupNames: string[] = [];
+  const logGroupLogicalIds = new Set<string>();
+  const template = stackArtifact.template as { [key: string]: any };
+
+  // do a first pass at identifying all log groups
+  if ('Resources' in template) {
+    for (const [key, resource] of Object.entries(template.Resources)) {
+      const definition = resource as { [key: string]: any };
+      if (definition.Type === 'AWS::Logs::LogGroup') {
+        logGroupLogicalIds.add(key);
+      } else if (Object.keys(RESOURCE_TYPES_WITH_IMPLICIT_LOGS).includes(definition.Type)) {
+        logicalIdsOfImplicitLogServices.set(key, definition.Type);
+      }
+    }
+  } else {
+    return {
+      env: resolvedEnv,
+      sdk,
+      logGroupNames: [],
+    };
+  }
+
+  // For each log group in the template make:
+  // 1. make sure it is not associated with an excluded type
+  // 2. see if it is associated with a resource type that implicitely
+  //   creates a log group so we know not to add the implicit log group later
+  for (const logGroupLogicalId of logGroupLogicalIds) {
+    const resourcesReferencingLogGroup = evaluateCfnTemplate.findReferencesTo(logGroupLogicalId);
+    for (const reference of resourcesReferencingLogGroup) {
+      if (IGNORE_LOGS_RESOURCE_TYPES.includes(reference.Type)) {
+        logGroupLogicalIds.delete(logGroupLogicalId);
+      } else if (Object.keys(RESOURCE_TYPES_WITH_IMPLICIT_LOGS).includes(reference.Type)) {
+        logicalIdsOfImplicitLogServices.delete(reference.LogicalId);
+      }
+    }
+  }
+
+  for (const logicalId of logGroupLogicalIds) {
+    const physicalNameInTemplate = getPhysicalNameProperty(template.Resources[logicalId], 'LogGroupName');
+    const groupName = await establishResourcePhysicalName(logicalId, physicalNameInTemplate, evaluateCfnTemplate);
+    if (groupName) {
+      logGroupNames.push(groupName);
+    }
+  }
+
+  // some resources can be created with a custom log group (handled above).
+  // if a custom log group is not created, then the service will create one with a
+  // specific name i.e. '/aws/codebuild/project-name'
+  for (const [logicalId, serviceName] of logicalIdsOfImplicitLogServices) {
```

Comment:
Sorry, I misunderstood what the values of this Map where. This should be:

```suggestion
  for (const [logicalId, cfnResourceType] of logicalIdsOfImplicitLogServices) {
```